### PR TITLE
Improvements to Tributary for CoffeeScript, Workflow, Error handling

### PR DESCRIPTION
**Pull request by EmbeddedMike**
There are a lot of changes in this pull request, and I apologize for not making this more granular. But here's the package:

1. The "execute" event is triggered many times during startup which can cause problem, for example when you self-reference the editor or reference a plugin before it's loaded. 
I set a flag __pluginsLoaded__ in inlet_handlebars to set this flag after the plugins are loaded, and guards in CoffeeContext for these two conditions. I think that loading plugins is the last initialization activity. If there was a way to suppress "execute" events until __pluginsLoaded__ is true, that would be better.

2. Fixed a bug in the inline-console plugin. It assumes that all widgets come from the same context (and thus same editor). Changed the plugin so that the clearLine function matches on line number and on cm editor. The call to clear all widgets was in the "onrender" callback. But that is nowhere in the code. I changed it to "pre:execute" 

3. Upgraded CoffeeScript to 1.9 both in package.json by copying the browser version into static/lib/coffee-script.js. This permits source maps and generators (with a browser that supports generators) among other things.

4. CoffeeScript compile error messages now appear in the CoffeeScript inlet editor. Originally done using inline-console's logJack function, but I changed it because error feedback wants to be in two parts a ^ under the starting token column, and the message itself.  __Plan to add CoffeeLint like you have done on JSLint__

5. Third argument (persist) added to console.logJack so you can have it add a line, rather than replace the current line.

6. Added console.logHTML so I could write two lines as a unit with arbitrary innerHTML, and a log-html CSS class with different coloring (error-like, for now) and without the indent and ">" __ I will remove the indent and ">" properties from the inline-console class CSS__

7. CoffeeScript runtime error messages now appear in the CoffeeScript inlet editor. The code for this has a bug which I will fix shortly. If the error is in a line in the inlet script, then the error is displayed at the right place. If the error is in a called routine, it is displayed on the first line in the inlet. It should display the error at the line where the call was done.

7. When using the CoffeeScript editor, If the word 'debugger' appears on its own line in the JS then sourceMaps are added to the compiled coffeescript. This lets you debug in coffeescript rather than JS. __This should really be done by messing with the tree, as you are doing with console.log in the plugin. __ Note: the reason I do it only when debugger is in the code is that the name of the file has to be changed each time, because Chrome (at least) treats each instance separately and does not get rid of old instances. 

8. `context` argument added to evaluated CoffeeScript function. This lets me modify some of the extension methods that I've created from within Tributary.

8. Improved workflow. Created shell script `runtributary` to watch the server and monitor files. If the server.js file or settings.js are changed, then the server is restarted. If any of the src files are changed then the current page is reloaded. More complete documentation is in `static/reload/README.md` I will add this information to the Wiki later.

### CODE REVIEW

**Comment by enjalot** [comment]
this is awesome! i look forward to merging in!

since we dont have tests i generally use representative inlets. do you have
any i can reference to try out the new features?

thanks!

**Comment by EmbeddedMike** [comment]
That's a great way to do it.
I will create some inlets today to cover the feature set.
Mike

**Comment by enjalot** [comment]
I went through your inlets:
http://tributary.io/s/6094415?user=EmbeddedMike
(fyi, if you create tributary inlets in localhost they wont show up in these production views because this uses the production database)

everything seemed to work fine with the pull request except this one:
http://tributary.io/inlet/34d645bc095fe4ca759c
which looks like the new version of coffeescript is a little more strict on some of the syntax?

anyway it looks pretty good to me, thanks so much for the contributions!

**Comment by enjalot** [comment]
I'm a little wary of including livereload in the production server though. is there a way to make sure it doesnt load then? or can we have a environment variable that turns it on locally when you run with ./runtributary?

**Comment by enjalot** [comment]
unfortunately i can't deploy it this weekend... I'm traveling and apparently this work laptop doesnt have the ssh key to get into the server =\
more time for you to make more thorough test inlets ;)

**Comment by EmbeddedMike** [comment]
Ian,
It should not be a problem, but if you want I will make the change. Here's
the story:

The code that I added to the server for livereload is:


livereload = require('livereload');
//Print debug messages, and watch 'md' files
server = livereload.createServer({debug: true, exts:['md']});
server.watch(__dirname + "/static");

That means that the livereload server watches the files in the static
directory for changes, and only then will send a message to the plugin
telling it to reload.

AFAIK (and you are the expert) the static directory is only changed during
the make process, and that only happens in development.

But even if there were changes, it would not do anything unless the
livereload plugin had been installed, and was enabled for a tributary page.

So to summarize: to livereload all three things have to happen:
1. Change file in static directory
2. Plugin installed by user
3. Plugin enabled by user, for that page

So I think program logic says 1 would not happen, and 2, and 3 are under
user control.

All that said, I will defer to you and if you want me to add an environment
variable or put it under switch control I will do that.
